Readme.md

# Component Search

  The component search application, powered by Express and Redis.

## API

### GET /all

  Respond with all packages:
  
  - `200` responds with json array of component.json contents

### GET /search/:query

  Search for `query`:
  
  - `500` shit exploded
  - `404` no packages, responds with json array `[]`
  - `200` packages found, responds with json array of component.json contents

# License

  MIT


History.md

0.3.3 / 2013-05-02 
==================

  * add stupid User-Agent to keep github happy

0.3.0 / 2012-11-10 
==================

  * add cors support
  * ugly fix for busted description

0.2.0 / 2012-10-29 
==================

  * add updating of "all" var

0.1.4 / 2012-10-29 
==================

  * add trimming of auth string

0.1.3 / 2012-10-27 
==================

  * change auth to Authorization header field

0.1.2 / 2012-10-27 
==================

  * add basic auth support
  * output res.status for errors from GH

0.1.1 / 2012-10-24 
==================

  * add error-handling
  * update wiki

0.1.0 / 2012-10-19 
==================

  * add hyphen support


util.js

/**
 * Parse words.
 * @param str {String}
 * @return {Array} Words.
 */

exports.words = function(str) {
  str = String(str).trim();
  if (!str) return [];
  return str.match(/\w+/).map(function(word){
    return word.toLowerCase();
  });
};

/**
 * Word keys from `words`.
 * @param words {Array}
 * @return {Array} Word keys.
 */

exports.wordKeys = function(words) {
  return words.map(function(str){
    return 'word:' + str;
  });
};

/**
 * Component keys from `names`.
 * @param names {Array}
 * @return {Array} Component keys.
 */

exports.componentKeys = function(names) {
  return names.map(function(str){
    return 'component:' + str;
  });
};

/**
 * List packages that depend on the supplied package, `pkg`.
 *
 * @param pkg
 * @param pkgs {Array} All packages to analyse.
 * @return {Array} Repos of packages that depend on `pkg`.
 */

exports.dependents = function(pkg, pkgs) {
  var repo = pkg.repo
  return pkgs.filter(blank).filter(function(pkg) {
    pkg.dependencies = pkg.dependencies || []
    return Object.keys(pkg.dependencies).indexOf(repo) !== -1
  }).map(function(pkg) {
    return pkg.repo
  })
}

/**
 * Check, if `pkg` is blank.
 *
 * @param pkg
 * @return {Boolean}
 */

function blank(pkg) {
  return !!pkg
}


fetch.js
/**
 * Module dependencies.
 */

var wiki = require('component-wiki')
  , request = require('superagent')
  , redis = require('redis')
  , db = redis.createClient()
  , util = require('./util')
  , fs = require('fs');

// NOTE: quick / horrid code lives here ;D

var pending = 0;
var packages;

removeSearchIndex();

function removeSearchIndex(){
  db.keys('word*', function(err, ids){
    if (err) throw err;
    if (!ids.length) return removeComponents();
    console.log('removing %d indexes', ids.length);
    db.del(ids, function(err){
      if (err) throw err;
      console.log('index removed')
      removeComponents();
    });
  });
}

function removeComponents(){
  db.keys('component*', function(err, ids){
    if (err) throw err;
    if (!ids.length) return fetch();
    console.log('removing %d components', ids.length);
    db.del(ids, function(err){
      if (err) throw err;
      console.log('removed components');
      fetch();
    });
  });
}

var auth = fs.readFileSync(process.env.HOME + '/.component-search-auth', 'ascii');
auth = new Buffer(auth.trim()).toString('base64');

function fetch() {
  var batch = wiki(function(err, pkgs){
    if (err) throw err;
    packages = pkgs;

    pkgs.forEach(function(pkg){
      if (!pkg) return;
      console.log('%s', pkg.name);

      var words = util.words(pkg.name);
      if (!pkg.description) console.log('"description" missing for %s', pkg.name);
      else words = words.concat(util.words(pkg.description));
      words = words.concat(pkg.keywords || []);

      pkg.dependents = util.dependents(pkg, pkgs);
      pkg.stars = 0;

      ++pending;
      request
      .get('https://api.github.com/repos/' + pkg.repo)
      .set('Authorization', 'Basic ' + auth)
      .set('User-Agent', 'Github')
      .end(function(res){
        done();

        if (res.ok) {
          pkg.stars = res.body.watchers_count;
          console.log('%s stars: %d', pkg.repo, pkg.stars);
        } else {
          console.log('github: %s %s', res.status, res.text);
        }

        ++pending;
        db.set('component:' + pkg.repo, JSON.stringify(pkg), done);
      })

      ++pending;
      db.sadd('components', pkg.repo, done);

      words.forEach(function(word){
        ++pending;
        db.sadd('word:' + word, pkg.repo, done);
      });
    });
  });

  batch.on('error', function(err){
    if (err.json) {
      console.error('invalid json: %s', err.url);
    } else {
      console.error(err.stack);
    }
  });
}

function done() {
  --pending || (function(){
    fs.writeFileSync(__dirname + '/components.json', JSON.stringify(packages));
    process.exit();
  })();
}



app.js

/**
 * Module dependencies.
 */

var express = require('express')
  , longjohn = require('longjohn')
  , fs = require('fs')
  , all = fs.readFileSync(__dirname + '/components.json')
  , redis = require('redis')
  , db = redis.createClient()
  , app = module.exports = express()
  , ms = require('ms')
  , util = require('./util');

// middleware

app.use(express.logger());
app.use(express.responseTime());
app.use(express.compress());

/**
 * Respond with packages.
 */

function reply(res) {
  return function reply(err, keys) {
    if (err) return res.send(500);
    if (!keys.length) return res.send(404, []);

    db.mget(util.componentKeys(keys), function(err, pkgs){
      if (err) return res.send(500);
      res.send(pkgs.map(JSON.parse));
    });
  }
}

/*
 * CORS support.
 */

app.all('*', function(req, res, next){
  res.set('Access-Control-Allow-Origin', '*');
  res.set('Access-Control-Allow-Methods', 'GET');
  res.set('Access-Control-Allow-Headers', 'X-Requested-With, Content-Type');
  if ('OPTIONS' == req.method) return res.send(200);
  next();
});

/**
 * GET all packages.
 */

app.get('/all', function(req, res){
  db.incr('stats:all');
  res.type('json');
  res.send(all);
});

/**
 * GET search :query.
 */

app.get('/search/:query', function(req, res){
  var query = util.words(req.params.query);

  // query stats
  db.incr('stats:queries');

  // word stats
  query.forEach(function(word){
    db.incr('stats:word:' + word);
  })

  // perform search
  query = util.wordKeys(query);
  db.sunion(query, reply(res));
});

/**
 * Update all.
 */

setInterval(function(){
  fs.readFile(__dirname + '/components.json', function(err, buf){
    if (err) return console.log(err.stack);
    console.log('updated all');
    all = buf;
  });
}, ms('5m'));
